zerver/lib/retention.py

```diff
@@ -56,42 +64,51 @@ def ids_list_to_sql_query_format(ids: List[int]) -> str:
 
     return ids_string
 
-def move_expired_messages_to_archive(realm: Realm) -> List[int]:
+def move_expired_messages_to_archive_by_recipient(recipient: Recipient,
+                                                  message_retention_days: int) -> List[int]:
     query = """
     INSERT INTO zerver_archivedmessage ({dst_fields}, archive_timestamp)
     SELECT {src_fields}, '{archive_timestamp}'
     FROM zerver_message
-    INNER JOIN zerver_userprofile ON zerver_message.sender_id = zerver_userprofile.id
     LEFT JOIN zerver_archivedmessage ON zerver_archivedmessage.id = zerver_message.id
-    WHERE zerver_userprofile.realm_id = {realm_id}
-          AND  zerver_message.pub_date < '{check_date}'
-          AND zerver_archivedmessage.id is NULL
+    WHERE zerver_message.recipient_id = {recipient_id}
+        AND zerver_message.pub_date < '{check_date}'
+        AND zerver_archivedmessage.id is NULL
     RETURNING id
     """
-    assert realm.message_retention_days is not None
-    check_date = timezone_now() - timedelta(days=realm.message_retention_days)
+    check_date = timezone_now() - timedelta(days=message_retention_days)
 
-    return move_expired_rows(Message, query, returning_id=True,
-                             realm_id=realm.id, check_date=check_date.isoformat())
+    return move_rows(Message, query, returning_id=True,
+                     recipient_id=recipient.id, check_date=check_date.isoformat())
 
-def move_expired_user_messages_to_archive(msg_ids: List[int]) -> List[int]:
-    if not msg_ids:
-        return []
+def move_expired_personal_and_huddle_messages_to_archive(realm: Realm) -> List[int]:
+    cross_realm_bot_ids_list = [get_user_including_cross_realm(email).id
+                                for email in settings.CROSS_REALM_BOT_EMAILS]
+    cross_realm_bot_ids = str(tuple(cross_realm_bot_ids_list))
+    recipient_types = (Recipient.PERSONAL, Recipient.HUDDLE)
 
+    # Archive expired personal and huddle Messages in the realm, except cross-realm messages:
     query = """
-    INSERT INTO zerver_archivedusermessage ({dst_fields}, archive_timestamp)
+    INSERT INTO zerver_archivedmessage ({dst_fields}, archive_timestamp)
     SELECT {src_fields}, '{archive_timestamp}'
-    FROM zerver_usermessage
-    LEFT JOIN zerver_archivedusermessage ON zerver_archivedusermessage.id = zerver_usermessage.id
-    WHERE zerver_usermessage.message_id IN {message_ids}
-        AND zerver_archivedusermessage.id is NULL
+    FROM zerver_message
+    INNER JOIN zerver_recipient ON zerver_recipient.id = zerver_message.recipient_id
+    INNER JOIN zerver_userprofile ON zerver_userprofile.id = zerver_message.sender_id
+    LEFT JOIN zerver_archivedmessage ON zerver_archivedmessage.id = zerver_message.id
+    WHERE zerver_userprofile.id NOT IN {cross_realm_bot_ids}
```

Comment:
I think you can remove this clause; cross-realm bots will always be in the "internal" realm in production (which has no other users in it), and so the realm_id query should filter them anyway.

In development ... we have for a long time planned to migrate cross-realm bots to live in an "internal" realm in the development environment too precisely because of this sort of consistency issue; if necessary we can pick up that project.  (I don't think it should be too much work; see https://github.com/zulip/zulip/issues/11015 for some basic notes).

Reply:
Removing it breaks the tests too though :( I added a TODO comment above this about removing the clause once #11015 is fixed - perhaps that's enough for now? I'll do that issue a bit later if needed, and remove the clause.

Reply:
Sounds good!